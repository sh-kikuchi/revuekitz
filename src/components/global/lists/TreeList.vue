<script setup lang="ts">
import { ref, computed } from 'vue'

interface Item {
  id: string
  title: string
  href?: string
}

interface Child {
  id: string
  title: string
  href?: string
  items: Item[]
}

interface Parent {
  id: string
  title: string
  href?: string
  children: Child[]
}

interface Grandparent {
  id: string
  title: string
  href?: string
  parents: Parent[]
}

const props = defineProps({
  id: {
    type: String,
    default: ''
  },
  class: {
    type: String,
    default: ''
  },
  style: {
    type: [String, Object],
    default: ''
  },
  styleReset: {
    type: Boolean,
    default: false
  },
  grandparents: {
    type: Array<Grandparent>,
    default: () => [
      {
        id: '1',
        title: 'grandparents #1',
        href: 'https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/Array/splice',
        parents: [
          {
            id: '1',
            title: 'parents #1',
            href: 'https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/Array/splice',
            children: [
              {
                id: '1',
                title: 'children #1',
                href: 'https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/Array/splice',
                items: [
                  {
                    id: '1',
                    title: 'Items #1',
                    href: 'https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/Array/splice'
                  },
                  {
                    id: '2',
                    title: 'Items #2',
                    href: 'https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/Array/splice'
                  }
                ]
              }
            ]
          }
        ]
      }
    ]
  }
})
const grandparants = ref(props.grandparents)

//binding classes
const bindingClass = computed(() => {
  if (props.styleReset == true) {
    return props.class
  } else {
    return `revuekitz-tree-list ${props.class} `
  }
})
</script>
<template>
  <div :id="props.id" :class="bindingClass" :style="props.style">
    <ul v-for="(grandparant, index) in grandparants" :key="index">
      <li>
        <a v-if="grandparant.hasOwnProperty('href')" :href="grandparant.href">{{
          grandparant.title
        }}</a>
        <ul v-for="(parent, index) in grandparant.parents" :key="index">
          <li>
            <a v-if="parent.hasOwnProperty('href')" :href="parent.href">{{ parent.title }}</a>
            <ul v-for="(child, index) in parent.children" :key="index">
              <li>
                <a v-if="child.hasOwnProperty('href')" :href="child.href">{{ child.title }}</a>
                <ul v-for="(item, index) in child.items" :key="index">
                  <li>
                    <a v-if="item.hasOwnProperty('href')" :href="item.href">{{ item.title }}</a>
                  </li>
                </ul>
              </li>
            </ul>
          </li>
        </ul>
      </li>
    </ul>
  </div>
</template>
<style scoped>
* {
  margin: 0;
  padding: 0;
  list-style: none;
}

.revuekitz-tree-list ul li ul {
  margin-top: 0.5em;
  margin-bottom: 0.75em;
  margin-left: 30px;
}

.revuekitz-tree-list ul li ul li {
  margin-bottom: 0.25em;
  position: relative;
}

.revuekitz-tree-list ul li ul li::before {
  content: '';
  position: absolute;
  top: -0.5em;
  left: -20px;
  width: 10px;
  height: calc(100% + 0.75em);
  border-left: 1px solid #3972b2;
}

.revuekitz-tree-list ul li ul li:last-child::before {
  height: calc(1em + 0.25em);
}

.revuekitz-tree-list ul li ul li::after {
  content: '';
  position: absolute;
  top: 0.75em;
  left: -20px;
  width: 10px;
  border-bottom: 1px solid #3972b2;
}
</style>
