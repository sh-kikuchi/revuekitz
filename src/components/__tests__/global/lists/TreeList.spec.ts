import { describe, it, expect } from 'vitest'
// TreeList.spec.ts
import { mount } from '@vue/test-utils'
import TreeList from '../../../../components/global/lists/TreeList.vue' // コンポーネントの実際のファイルパスに置き換える

describe('TreeList', () => {
  it('renders grandparents, parents, children, and items with correct links', async () => {
    const grandparents = [
      {
        id: '1',
        title: 'Grandparent 1',
        href: 'https://example.com',
        parents: [
          {
            id: '1',
            title: 'Parent 1',
            href: 'https://example.com',
            children: [
              {
                id: '1',
                title: 'Child 1',
                href: 'https://example.com',
                items: [
                  {
                    id: '1',
                    title: 'Item 1',
                    href: 'https://example.com'
                  },
                  {
                    id: '2',
                    title: 'Item 2',
                    href: 'https://example.com'
                  }
                ]
              }
            ]
          }
        ]
      }
    ]

    const wrapper = mount(TreeList, {
      props: {
        grandparents
      }
    })

    // Check if the component renders grandparents correctly
    const grandparentLink = await wrapper.find('a')
    expect(grandparentLink.attributes('href')).toBe('https://example.com')
    expect(grandparentLink.text()).toBe('Grandparent 1')

    // Check if it renders parents correctly
    const parentLink = await wrapper.find('ul > li > ul > li > a')
    expect(parentLink.attributes('href')).toBe('https://example.com')
    expect(parentLink.text()).toBe('Parent 1')

    // Check if it renders children correctly
    const childLink = await wrapper.find('ul > li > ul > li > ul > li > a')
    expect(childLink.attributes('href')).toBe('https://example.com')
    expect(childLink.text()).toBe('Child 1')

    // Check if it renders items correctly
    const itemLink = await wrapper.find('ul > li > ul > li > ul > li > ul > li > a')
    expect(itemLink.attributes('href')).toBe('https://example.com')
    expect(itemLink.text()).toBe('Item 1')

    // Additional checks for other items, attributes, and structure can be added as needed
  })
})
