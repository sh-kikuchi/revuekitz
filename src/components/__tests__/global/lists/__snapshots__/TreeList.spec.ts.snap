// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TreeList > renders correct HTML structure based on props 1`] = `
"<div data-v-0dc24891="" class="treelist">
  <ul data-v-0dc24891="">
    <li data-v-0dc24891=""><a data-v-0dc24891="" href="https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/Array/splice">Parents #1</a>
      <ul data-v-0dc24891="">
        <li data-v-0dc24891=""><a data-v-0dc24891="" href="https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/Array/splice">Children #1</a>
          <ul data-v-0dc24891="">
            <li data-v-0dc24891=""><a data-v-0dc24891="" href="https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/Array/splice">GrandChildren #1</a>
              <ul data-v-0dc24891="">
                <li data-v-0dc24891=""><a data-v-0dc24891="" href="https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/Array/splice">Items #1</a></li>
              </ul>
            </li>
          </ul>
        </li>
      </ul>
    </li>
  </ul>
</div>"
`;
